DAOS-15338 dfuse: Configure fuse to allow writes to populate cache.

ashleypittman opened the pull request:
Change dfuse data caching settings to allow writes to populate cache.
Track the data cache validity separately from attribute cache validity to
allow finer grained control.

On close of a file then speculatively call getattr to get updated size/mtime
information and use this as a reference for future dcache invalidation.

For getattr calls after close then use the pre-loaded stat data leading to
increased performance by moving the getattr off the critical path (only for
files which were written to, files which aren't do not invalidate the atttr cache).

Use the FUSE_CAP_EXPLICIT_INVAL_DATA kernel feature to instruct the
kernel not to automatically invalidate data in the size/mtime has changed, this
is now performed by dfuse.

Overall this is a performance improvement for stat-after-close but the major
change is that a write/read workflow will now correctly use the kernel page
cache.  Previously writes were modifying mtime values and that was incorrectly
triggering the kernel to discard it's cache unnecessarily.

Add testing of write/read and read/read I/O patterns verified by the dfuse
statistics that data caching is working as expected.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
DAOS-15338 dfuse: add a test for read caching.

Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/1/execution/node/353/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/1/execution/node/346/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/1/execution/node/314/log

ashleypittman committed:
Check for flag before logging it.

Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/2/execution/node/354/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/2/execution/node/332/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/2/execution/node/327/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/2/execution/node/270/log

ashleypittman committed:
Add a ftest for this.

Test-tag: DFuseReadTest
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false

Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix the test tags.

Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/3/execution/node/740/log

ashleypittman committed:
Try loading dfuse before access.

Skip-build-ubuntu20-rpm: true

Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/3/display/redirect

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/4/execution/node/653/log

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/4/display/redirect

ashleypittman committed:
Setup a container.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 9 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/5/testReport/

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/5/display/redirect

ashleypittman committed:
Change the pool setup.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/6/display/redirect

ashleypittman committed:
Turn on dfuse debugging.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/7/display/redirect

ashleypittman committed:
Use a larger write size.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: true
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/8/execution/node/324/log

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/9/display/redirect

ashleypittman committed:
Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/10/execution/node/751/log

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/10/execution/node/767/log

ashleypittman committed:
Debug the test.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix spelling.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/11/execution/node/747/log

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/11/execution/node/722/log

ashleypittman committed:
Re-attach the data before parsing it.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
check the number of reads.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/13/execution/node/751/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/13/execution/node/767/log

ashleypittman committed:
Fix json key.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/14/testReport/

ashleypittman committed:
Make new code into a dfuse method.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/15/testReport/

ashleypittman committed:
Tidy up.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/16/testReport/

ashleypittman committed:
Remove date checks.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Required-githooks: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Start adding back in the eviction code.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Further tweak build

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/20/testReport/

ashleypittman committed:
Save my changes.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/23/testReport/

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/24/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
PR-repos: fuse@PR-11

ashleypittman committed:
Hotpatch Jenkinsfile.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
PR-repos: fuse3@PR-11

Skip-build-leap15-rpm: true
Skip-build-el9-rpm: true
Skip-func-test-el9: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/25/execution/node/319/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/25/execution/node/345/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/25/execution/node/277/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/27/execution/node/332/log

ashleypittman committed:
Enpty commit for testing.

Features: dfuse
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
PR-repos: fuse3@PR-11
Skip-build-leap15-rpm: true
Skip-build-el9-rpm: true
Skip-func-test-el9: true
Skip-build-ubuntu20-rpm: true

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/29/execution/node/620/log

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Features: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/31/testReport/

ashleypittman committed:
Drop the data cache if metadata has changesd.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-tests: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix lint.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Silence a new warning.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest DaosBuild
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Back out NLT metadata changes.

Skip-build-ubuntu20-rpm: true
Test-tag: DFuseReadTest DaosBuild
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/37/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/38/testReport/

ashleypittman committed:
Reply with the new inode data.

Test-tag: dfuseF

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Only use the cached value once.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

ashleypittman committed:
Fir clang-format and run tests.

Skip-build-ubuntu20-rpm: true
Test-tag: dfuse
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/41/testReport/

ashleypittman committed:
Run cachning test

Skip-build-ubuntu20-rpm: true
Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-tests: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false

Merge branch 'master' into amd/dfuse-read-cache

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/43/testReport/

ashleypittman committed:
Add debugging to test.

Skip-build-ubuntu20-rpm: true
Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-tests: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/45/execution/node/852/log

ashleypittman committed:
Fix caching check test.

Skip-build-ubuntu20-rpm: true
Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-tests: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/47/testReport/

ashleypittman committed:
Use the new fuse version.

PR-repos: fuse3@PR-11
Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-tests: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-build-leap15-rpm: true
Skip-build-el9-rpm: true
Skip-func-test-el9: false
Skip-build-ubuntu20-rpm: true

Skip-func-hw-test: true
Skip-func-test: true
Quick-Functional: true
Test-tag: dfuse

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/49/execution/node/229/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/51/execution/node/229/log

daosbuild1 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-13927/52/display/redirect

ashleypittman committed:
Bump timeouts.

Skip-build-ubuntu20-rpm: true
Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-scan-el8-rpms: true
Skip-unit-test: true
Skip-unit-tests: true
Skip-unit-test-memcheck: true
Skip-build-el8-rpm: false
Skip-func-test-el9: false

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/54/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Features: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Do not run the complex rename test with caching.

Features: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Features: dfuse
Priority: 2

ashleypittman committed:
Fix flake.

Features: dfuse
Priority: 2
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Minor changes.

Features: dfuse
Priority: 2
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/59/execution/node/1406/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/59/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Test-tag: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Reset default cache settings.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/62/execution/node/1400/log

ashleypittman committed:
DAOS-15745 dfuse: Add the pre_read metrics whilst holding reference.

Increase the pre-read statistics before replying to the read,
otherwise the oh might not be valid which can lead to unexpected
behaviour.

Test-tag: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/64/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache and run with new libfuse.

Test-tag: dfuse

Skip-func-test-el9: false
PR-repos: fuse3@PR-11
Skip-fault-injection-test: true
Skip-unit-tests: true
Skip-build-ubuntu20-rpm: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/65/execution/node/292/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/65/execution/node/343/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/65/execution/node/352/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/65/execution/node/353/log

daosbuild1 commented:
Test stage Functional on EL 9 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/66/testReport/

ashleypittman committed:
Tune down a debug message.

Test-tag: dfuse,-test_dfuse_find
Skip-fault-injection-test: true
Skip-unit-tests: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix a bug where reads weren't going into cache.

Improve test to catch this.

zzTest-tag: dfuse,-test_dfuse_find
Test-tag: test_dfuse_read
Skip-fault-injection-test: true
Skip-unit-tests: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix flake and run more tests.

Test-tag: dfuse,-test_dfuse_find
Skip-fault-injection-test: true
Skip-unit-tests: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daltonbohning commented on src/tests/ftest/dfuse/read.py:
```diff
@@ -96,3 +96,98 @@ def test_dfuse_pre_read(self):
+
+        self.dfuse.disable_wb_cache.value = True
+
+        self.dfuse.env["D_LOG_MASK"] = "INFO,DFUSE=DEBUG"
+        self.dfuse.env["DD_MASK"] = "ALL"
+        self.dfuse.env["DD_SUBSYS"] = "ALL"
```
Do you want this set for both test cases?
If so, we can put this in the config
```
dfuse:
  disable_wb_cache: true
  env_vars:
    - D_LOG_MASK=INFO,DFUSE=DEBUG
    - DD_MASK=ALL
    - DD_SUBSYS=ALL
```


And if it needs to be different, I'd recommend different namespaces like
```
dfuse:
  disable_wb_cache: true
dfuse_debug:
  disable_wb_cache: true
  env_vars:
    - D_LOG_MASK=INFO,DFUSE=DEBUG
    - DD_MASK=ALL
    - DD_SUBSYS=ALL
```

And then initialize like
```
self.load_dfuse(self.hostlist_clients, '/run/dfuse/*')
self.load_dfuse(self.hostlist_clients, '/run/dfuse_debug/*')
```

daltonbohning commented on src/tests/ftest/dfuse/read.py:
```diff
@@ -96,3 +96,98 @@ def test_dfuse_pre_read(self):
+        read_calls = data["statistics"].get("read", 0)
+        write_calls = data["statistics"].get("write")
+
+        print(f"Test caused {write_calls} write and {read_calls} reads calls")
```
```suggestion
        self.log.info("Test caused %d write and %d reads calls", write_calls, read_calls)
```

mchaarawi commented on src/client/dfuse/ops/open.c:
```diff
@@ -203,16 +304,22 @@ dfuse_cb_release(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi)
+		dfuse_dcache_set_time(oh->doh_ie);
+		if (oc == 1)
+			/* Fire off a ostatx call to fetch the new mtime */
+			size_resample(dfuse_info, oh->doh_ie);
```
a stat on a file close is going to be pretty bad for workloads that have millions of files ..

ashleypittman replied:
This is only for files which were written to.  If there's any access to the file afterwards then the kernel would do this on next access anyway so the only additional network load here is for files which are written to but then not accessed again.

mchaarawi replied:
most of workloads we have seen today write/read to the file with 1 handle. i.e. there is no close and re-access.
What workload are you thinking about here that does this write, close, open, read that would benefit from this?

wiliamhuang replied:
Most applications do not read the files in a short time after writing files. Some applications (e.g., computation chemistry) may do write & close then open & read later because intermediate data are too large to fit in memory.

ashleypittman replied:
You're right this will help with some I/O patterns and not others, I don't think we can speculate accurately on which is more common but the bug is clear - with current master the cache is dropped on open in the open/write/close -> open/read/close case which is a simple bug due to having the kernel handle cache validity decisions based on incorrect assumptions on how mtime updates are handled.

Some applications and workflows won't be triggering this bug and will therefore be unaffected by it, a lot more will be triggering it and benefit from the fix, interactive workloads will probably benefit the most but plenty of HPC ones will.  IOR certainly would if using the `-C`/reorder flag and the right rank allocation (although this option is mainly there to force clients onto other nodes to avoid the use of in-node caching).

mchaarawi replied:
my main concern is the effect of an extra stat on those unaffected workloads. i do not agree that they are "unaffected".
anyway since you mentioned IOR is a use case that can benefit, let's gather some results then before we decide on this fix.

ashleypittman replied:
I did have some numbers on this when I was developing it and it was showing a very high proportion of these speculative stat calls were returned to the kernel - but it'll be entirely application dependant as you say.

Anyway - the extra stat is asynchronous but launched on close so it'll be impossible to measure from a client perspective as the extra cost, if it is indeed extra, is all server side so won't show up in any simple benchmark.

wiliamhuang commented on src/client/dfuse/dfuse_core.c:
```diff
@@ -1048,6 +1064,51 @@ dfuse_mcache_get_valid(struct dfuse_inode_entry *ie, double max_age, double *tim
+	struct timespec left;
+	double          time_left;
+
+	D_ASSERT(max_age != -1);
+	D_ASSERT(max_age >= 0);
```
Minor issue, if "max >= 0", then the first assertion always is true. Do we need the first assertion?

ashleypittman replied:
True, we don't.  I've used this check elsewhere in the code though, for some cache types (data) then -1 is a valid cache age so the reason for having two checks here is that if it does fail we know why, was the cache incorrectly initialised/loaded in which case it'll be -1 or was a specific age incorrectly calculated in which case it's be a different negative value.

wiliamhuang replied:
I agree. Thank you for your clarification!

wiliamhuang commented on src/client/dfuse/dfuse_core.c:
```diff
@@ -1048,6 +1064,51 @@ dfuse_mcache_get_valid(struct dfuse_inode_entry *ie, double max_age, double *tim
 	return use;
 }
 
+/* Set a timer to mark cache entry as valid */
```
minor: "timer"-->"time"?

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Features: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman commented on src/client/dfuse/dfuse.h:
```diff
@@ -998,6 +1000,37 @@ struct dfuse_inode_entry {
 	/** File has been unlinked from daos */
 	bool                      ie_unlinked;
 
+	/* Data cache metadata, list known size/mtime for file, if these have been updated then
```
This is the information used to determine if the data cache is to be considered valid - it's data about the data cache.

mchaarawi replied:
yea maybe reword to something like "metadata for the data-cache"

ashleypittman replied:
I can re-push with that, any other changes whilst I'm at it?

mchaarawi replied:
this is just a minor comment that does not require a repush. there is a bigger concern on the stat after close that is more fundamental and needs to be discuss more before any repush..

ashleypittman committed:
Tidy up test code.

Test-tag: test_dfuse_caching_check
Skip-fault-injection-test: true
Skip-unit-tests: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Run with features.

Features: dfuse,-test_dfuse_daos_build_wt_pil4dfs

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/72/execution/node/1436/log

knard-intel commented on src/client/dfuse/ops/open.c:
```diff
@@ -132,6 +135,101 @@ dfuse_cb_open(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi)
+
+	sem_post(&eqt->de_sem);
+
+	return;
```
In case of success, the resources such as `ev` are not freed ?

knard38 reviewed (commented):
Mostly LGTM for what I understand.
Some points to clarify as I am not used to this part of the code.

mchaarawi reviewed (changes requested):
as discussed in WG, the stat on close is pretty significant hit performance wise and to the DAOS cluster itself in case of large deployments and widely striped files

mchaarawi commented:
> @mchaarawi having the extra asynchronous getattr on close will be a big win for many of the first impression applications (such as when people try to clone and compiler in a dfuse mount). In such cases, fuse will do subsequent getattr anyway, this just caches it preemptively and, more importantly, keeps the write cache for the subsequent reads. For performance sensitive cases, I suspect the extra asynchronous stat is somewhat negligible compared to the expected normal overhead of dfuse. It already does a million getattr's.

im pretty sure that this does server some use cases. im worried though about use cases that are just now doing an extra stat on close.. say an app that just writes a file and never accesses it again till much later. the stat in this case of widely stripped files if there are a bunch of them, would be bad. so yes this probably improves 1 use case but can be problemetic in others

mchaarawi commented:
> Actually, maybe we can augment dfs_write to return the max mtime for the chunks it writes? And we can approximate the size based on what we wrote and what we already know?

the io path should never be modified IMO to do extra metadata things like this.

jolivier23 commented:
> > @mchaarawi having the extra asynchronous getattr on close will be a big win for many of the first impression applications (such as when people try to clone and compiler in a dfuse mount). In such cases, fuse will do subsequent getattr anyway, this just caches it preemptively and, more importantly, keeps the write cache for the subsequent reads. For performance sensitive cases, I suspect the extra asynchronous stat is somewhat negligible compared to the expected normal overhead of dfuse. It already does a million getattr's.
> 
> 
> 
> im pretty sure that this does serve some use cases. im worried though about use cases that are just now doing an extra stat on close.. say an app that just writes a file and never accesses it again till much later. the stat in this case of widely stripped files if there are a bunch of them, would be bad. so yes this probably improves 1 use case but can be problemetic in others

Along the lines of my next comment, what if we locally tracked size and mtime on the client object and allowed the application to query the values.  This would be a good approximation in the cases I'm talking about and just get invalidated for others.

mchaarawi commented:
> > > @mchaarawi having the extra asynchronous getattr on close will be a big win for many of the first impression applications (such as when people try to clone and compiler in a dfuse mount). In such cases, fuse will do subsequent getattr anyway, this just caches it preemptively and, more importantly, keeps the write cache for the subsequent reads. For performance sensitive cases, I suspect the extra asynchronous stat is somewhat negligible compared to the expected normal overhead of dfuse. It already does a million getattr's.
> > 
> > 
> > im pretty sure that this does serve some use cases. im worried though about use cases that are just now doing an extra stat on close.. say an app that just writes a file and never accesses it again till much later. the stat in this case of widely stripped files if there are a bunch of them, would be bad. so yes this probably improves 1 use case but can be problemetic in others
> 
> Along the lines of my next comment, what if we locally tracked size and mtime on the client object and allowed the application to query the values. This would be a good approximation in the cases I'm talking about and just get invalidated for others.
> 
> In other words the local object would keep track of locally highest offset and mtime

the mtime we use today there is based on server time stamps (the max epoch thing). so if dfuse wants to use it's own timestamps to do this, there might be issues of client and servers are not properly synchronized.
but in a single case dfuse instance, it might not be a big deal?
we cannot do his at the dfs level because it we currently do not do any caching there. but we might do that in the future.

ashleypittman committed:
Merge branch 'master' into amd/dfuse-read-cache

Test-tag: dfuse

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/317/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/314/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/273/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/453/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/357/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13927/73/execution/node/519/log

ashleypittman committed:
Fix merge.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Bump test runtime.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Update spelling.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13927/75/testReport/
